Mage.Sets/src/mage/cards/w/WaytaTrainerProdigy.java

```diff
@@ -0,0 +1,162 @@
+        super(card);
+    }
+
+    @Override
+    public WaytaTrainerProdigy copy() {
+        return new WaytaTrainerProdigy(this);
+    }
+}
+
+// Based on Crown of Gondor and SourceTargetsPermanentCondition
+enum WaytaTrainerProdigyAdjuster implements CostAdjuster {
+    instance;
+
+    private static final FilterPermanent filter
+            = new FilterControlledCreaturePermanent("a creature you control");
+
+    @Override
+    public void adjustCosts(Ability ability, Game game) {
+        Target secondTarget = null;
+        for (Target target : ability.getTargets()){
+            if (target.getTargetTag() == 2){
+                secondTarget = target;
+                break;
+            }
+        }
+        if (secondTarget == null){
+            return;
+        }
+        // Having to call getFirstTarget() on a Target object called secondTarget
+        // (because it's the second target of a two-target ability)
+        // seems like an insult, but this is just getting the UUID of that target
+        Permanent permanent = game.getPermanentOrLKIBattlefield(secondTarget.getFirstTarget());
+        if (permanent != null && filter.match(permanent, ability.getControllerId(), ability, game)) {
+            CardUtil.reduceCost(ability, 2);
+        }
+    }
+}
+
+class WaytaTrainerProdigyEffect extends ReplacementEffectImpl {
+
+    WaytaTrainerProdigyEffect() {
+        super(Duration.WhileOnBattlefield, Outcome.Benefit);
+        staticText = "If a creature you control being dealt damage causes a triggered ability " +
+                "of a permanent you control to trigger, that ability triggers an additional time";
+    }
+
+    private WaytaTrainerProdigyEffect(final WaytaTrainerProdigyEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public WaytaTrainerProdigyEffect copy() {
+        return new WaytaTrainerProdigyEffect(this);
+    }
+
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.NUMBER_OF_TRIGGERS;
+    }
+
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+        NumberOfTriggersEvent numberOfTriggersEvent = (NumberOfTriggersEvent) event;
+        Permanent sourcePermanent = game.getPermanent(numberOfTriggersEvent.getSourceId());
+        if (sourcePermanent == null || !sourcePermanent.isControlledBy(source.getControllerId())) {
+            return false;
+        }
+
+        GameEvent sourceEvent = numberOfTriggersEvent.getSourceEvent();
+        if (sourceEvent == null) {
+            return false;
+        }
+
+        switch (sourceEvent.getType()) {
+            case DAMAGED_BATCH_FOR_PERMANENTS:
+            case DAMAGE_PERMANENT:
+            case DAMAGED_PERMANENT:
+            case COMBAT_DAMAGE_STEP:
+            case COMBAT_DAMAGE_STEP_PRE:
+            case COMBAT_DAMAGE_STEP_PRIORITY:
+            case COMBAT_DAMAGE_STEP_POST:
+                return true;
```

Comment:
Yeah, this list of events looks a bit questionable, though superfluous inclusions shouldn't cause any bugs, so not a big deal.

FYI- convention is that events come in pairs. The present tense (e.g. `DAMAGE_PERMANENT`) is for checking against replacement effects before the thing (damage) happens. The past tense (e.g. `DAMAGED_PERMANENT`) is what is fired after the thing (damage) happens and is checked by watchers/triggers.
The `STEP` events are really only for "At..." triggers, or for watchers to clear/update their state- shouldn't be related to damage dealt.

Reply:
Yeah if this was just something that causes its own effect on a damage, i would just pick one event and go with that, but this card is in the awkward position of having to check for triggers that use any event that deals with a creature being dealt damage. So if in the future some card uses the COMBAT_DAMAGE_STEP_PRE enum for its trigger of some effect (example chosen because it looks like no cards use that to do a trigger that could be retriggered), then this card has to look for those too. Is there a standard that new cards should need to conform to about this that i could narrow down this list?

Reply:
1. NUMBER_OF_TRIGGERS process any non state base triggers. "Dealt damage" triggers by DAMAGED_PERMANENT and DAMAGED_BATCH_FOR_PERMANENTS events only. So you need only it, not all that weird pre and post steps.
2. Batch event must be process as `BatchGameEvent` with `getTargets` usage. `numberOfTriggersEvent.getSourceId()` will not work for such cards (for cards that triggers by DAMAGED_BATCH_FOR_PERMANENTS).
3. `numberOfTriggersEvent.getSourceId()` smells bad cause rules must check BOTH permanents:
  a. Creature you control being dealt damage
  b. Permanent you control to trigger

Reply:
4. Add official rules and make sure it works:
> Wayta, Trainer Prodigy's last ability affects only triggered abilities whose trigger conditions refer specifically to damage being dealt, such as the ability granted by Mephidross Vampire or the last ability of Wrathful Raptors. It does not affect triggered abilities that would trigger because of the results of that damage. For example, if you control an Ajani's Pridemate ("Whenever you gain life, put a +1/+1 counter on Ajani's Pridemate") and you activate Wayta, Trainer Prodigy's second ability targeting a creature you control with lifelink and another creature you control, the triggered ability of Ajani's Pridemate will still trigger only once. 
(2023-11-10)

5. Make sure it support multiple copies:
> If you somehow control two copies of Wayta, Trainer Prodigy, a creature you control being dealt damage causes abilities to trigger three times, not four. A third Wayta, Trainer Prodigy causes abilities to trigger four times, a fourth causes abilities to trigger five times, and so on.
(2023-11-10)

Reply:
> 2. Batch event must be process as BatchGameEvent with getEvents::getSourceId usage. numberOfTriggersEvent.getSourceId() will not work for such cards (for cards that triggers by DAMAGED_BATCH_FOR_PERMANENTS).

This is to say that the `GameEvent`should be cast to a `BatchGameEvent`, and the source ID be retrieved from `getSourceID`?

If this issue is true, then [Isshin two heavens as one](https://scryfall.com/card/neo/224/isshin-two-heavens-as-one) and 
[Wulfgar of Icewind Dale](https://scryfall.com/card/afc/56/wulfgar-of-icewind-dale) are probably worth a second look--though, their triggers are on "A creature [you control] _attacking_", so maybe they need to play by different rules with this respect, but the way `applies` works for Wayta was taken directly from Isshin.

Reply:
> 3\. `numberOfTriggersEvent.getSourceId()` smells bad cause rules must check BOTH permanents:
>         a. Creature you control being dealt damage
>         b. Permanent you control to trigger

I'm not sure how to check for this, since i dont see any other examples of this kind of effect being implemented except for the aforementioned Isshin and Wulfgar, and those cards don't seem to check for this.

Reply:
Yes. BatchEvent will raise game error (wrong usage exception) on direct getSourceId usage. So you must search events list instead for any “compatible/controlled” sourceId.

Reply:
> 7\. Add official rules and make sure it works:
> 
> 
> > Wayta, Trainer Prodigy's last ability affects only triggered abilities whose trigger conditions refer specifically to damage being dealt, such as the ability granted by Mephidross Vampire or the last ability of Wrathful Raptors. It does not affect triggered abilities that would trigger because of the results of that damage. For example, if you control an Ajani's Pridemate ("Whenever you gain life, put a +1/+1 counter on Ajani's Pridemate") and you activate Wayta, Trainer Prodigy's second ability targeting a creature you control with lifelink and another creature you control, the triggered ability of Ajani's Pridemate will still trigger only once.
> > (2023-11-10)

Tested this exact case and it behaves as described. by "Add official rules" do you mean i should ass this text somewhere in the java class?

Reply:
> do you mean i should ass this text somewhere in the java class

Yes, it's a good style to add card related rules to the class

Reply:
> 5\. Make sure it support multiple copies:
> 
> 
> > If you somehow control two copies of Wayta, Trainer Prodigy, a creature you control being dealt damage causes abilities to trigger three times, not four. A third Wayta, Trainer Prodigy causes abilities to trigger four times, a fourth causes abilities to trigger five times, and so on.
> > (2023-11-10)

I confirmed that the retriggers scale linearly as described, but i ran into a weird bug--I copied Wayta using [Double Major](https://scryfall.com/card/stx/179/double-major) so that it wouldnt be legendary, and the nonlegendary card on the battlefield has the expected test and info, but its tap ability does nothing--it lets me select two creatures, the cost adjuster works, and then they just dont fight and nothing is logged.

What might be causing that?

Reply:
> > do you mean i should ass this text somewhere in the java class
> 
> Yes, it's a good style to add card related rules to the class

Noted--sorry for that typo lmao

Reply:
> I'm not sure how to check for this

Code example:
``` java
        NumberOfTriggersEvent numberOfTriggersEvent = (NumberOfTriggersEvent) event;
        GameEvent sourceEvent = numberOfTriggersEvent.getSourceEvent();


        // permanent you control to trigger
        Permanent triggeredPermanent = game.getPermanentOrLKIBattlefield(numberOfTriggersEvent.getSourceId());
        if (triggeredPermanent == null || !triggeredPermanent.isControlledBy(source.getControllerId())) {
            return false;
        }

        // creature you control being dealt damage
        List<UUID> damagedPermanents = new ArrayList<>();
        if (sourceEvent instanceof BatchGameEvent) {
            damagedPermanents.addAll(((BatchGameEvent) sourceEvent).getTargets());
        } else {
            damagedPermanents.add(sourceEvent.getTargetId());
        }
        if (damagedPermanents
                .stream()
                .map(game::getPermanentOrLKIBattlefield)
                .filter(Objects::nonNull)
                .noneMatch(p -> p.isControlledBy(source.getControllerId()))) {
            return false;
        }

        // can trigger additional time
        // ...
```

Reply:
> it lets me select two creatures, the cost adjuster works, and then they just dont fight and nothing is logged.

If no logs and you can't continue to play then it can be an AI freeze. Just wait some time to continue.

Reply:
If you run in debug mode (IntelliJ IDEA) then you can press dump thread button and check green/active threads -- green "AI/SIM" for AI calculation, "CALL" for user command calculations and "GAME" for game calculations:

![shot_240106_001558](https://github.com/magefree/mage/assets/8344157/813f41bd-bc41-4d41-8db0-3a74048472c7)

Reply:
> If no logs and you can't continue to play then it can be an AI freeze. Just wait some time to continue.

The AI appears to have responded fine--it moves me to the generic "Play spells and abilities" stage of the step. the fight just didnt happen for some reason.

Log follows. at the end you can see the skip attack log being the first log item to show up after two creatures were supposed to kill each other in precombat main.
```
3:18 PM, T0: Human won the toss
3:18 PM, T0: Human: tried to apply cheat commands
3:18 PM, T0: Human chooses that they take the first turn
3:18 PM, T0: Human draws seven cards
3:18 PM, T0: Computer draws seven cards
3:18 PM, T0: Human keeps hand
3:18 PM, T0: Computer decides to take mulligan
3:18 PM, T0: Computer's library is shuffled
3:18 PM, T0: Computer mulligans for free.
3:18 PM, T0: Computer draws seven cards
3:18 PM, T0: Computer decides to take mulligan
3:18 PM, T0: Computer's library is shuffled
3:18 PM, T0: Computer mulligans for free.
3:18 PM, T0: Computer draws seven cards
3:18 PM, T0: Computer decides to take mulligan
3:18 PM, T0: Computer's library is shuffled
3:18 PM, T0: Computer mulligans down to 6 cards
3:18 PM, T0: Computer draws seven cards
3:18 PM, T0: Computer puts a card from hand to the bottom of their library
3:18 PM, T0: Computer decides to take mulligan
3:18 PM, T0: Computer's library is shuffled
3:18 PM, T0: Computer mulligans down to 5 cards
3:18 PM, T0: Computer draws seven cards
3:18 PM, T0: Computer puts a card from hand to the bottom of their library
3:18 PM, T0: Computer puts a card from hand to the bottom of their library
3:18 PM, T0: Computer keeps hand
3:18 PM: Turn 1 Human (40 - 40)
3:18 PM, T1.DR: Human draws a card
3:18 PM, T1.M1: Human: tried to apply cheat commands
3:18 PM, T1.M1: Human: tried to apply cheat commands
3:18 PM, T1.M1: Human: tried to apply cheat commands
3:18 PM, T1.M1: Human: tried to apply cheat commands
3:18 PM, T1.M1: Human casts [Wayta, Trainer Prodigy](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Wayta,%20Trainer%20Prodigy) [092]
3:18 PM, T1.M1: Human casts [Double Major](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Double%20Major) [461] targeting [092]
3:18 PM, T1.M1: Human created a copy of Wayta, Trainer Prodigy [092]
3:18 PM, T1.M1: Human puts [Double Major](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Double%20Major) [461] from stack into their graveyard
3:18 PM, T1.M1: [Wayta, Trainer Prodigy](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Wayta,%20Trainer%20Prodigy) [d86] enters the battlefield as a token under Human's control'
3:18 PM, T1.M1: Human puts [Wayta, Trainer Prodigy](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Wayta,%20Trainer%20Prodigy) [092] from stack onto the Battlefield
3:18 PM, T1.M1: Human casts [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5]
3:18 PM, T1.M1: Human puts [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5] from stack onto the Battlefield
3:18 PM, T1.M1: Human casts [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937]
3:18 PM, T1.M1: Human puts [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937] from stack onto the Battlefield
3:19 PM, T1.M1: Human activates: target creature you control fights another target creature. This ability costs 2 less to activate if it targets two creatures you control. from [Wayta, Trainer Prodigy](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Wayta,%20Trainer%20Prodigy) [092] targeting [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5][Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937]
3:19 PM, T1.M1: [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937] deals 4 damage to [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5]
3:19 PM, T1.M1: [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5] deals 3 damage to [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937]
3:19 PM, T1.M1: Human - Ability triggers: [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937] - Enrage — Whenever [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937] is dealt damage, draw a card.
3:19 PM, T1.M1: Human - Ability triggers: [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937] - Enrage — Whenever [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937] is dealt damage, draw a card.
3:19 PM, T1.M1: Human - Ability triggers: [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937] - Enrage — Whenever [Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937] is dealt damage, draw a card.
3:19 PM, T1.M1: Human - Ability triggers: [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5] - Enrage — Whenever [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5] is dealt damage, you gain 2 life.
3:19 PM, T1.M1: Human - Ability triggers: [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5] - Enrage — Whenever [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5] is dealt damage, you gain 2 life.
3:19 PM, T1.M1: Human - Ability triggers: [eb5] - Enrage — Whenever [eb5] is dealt damage, you gain 2 life.
3:19 PM, T1.M1: Human gains 2 life (source: [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5])
3:19 PM, T1.M1: Human gains 2 life (source: [eb5])
3:19 PM, T1.M1: Human gains 2 life (source: [eb5])
3:19 PM, T1.M1: Human draws a card
3:19 PM, T1.M1: Human draws a card
3:19 PM, T1.M1: Human draws a card
3:19 PM, T1.M1: Human puts [3e5] from hand onto the Battlefield
3:19 PM, T1.M1: Human plays [Forest](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Forest) [3e5]
3:19 PM, T1.M1: Human activates: target creature you control fights another target creature. This ability costs 2 less to activate if it targets two creatures you control. from [Wayta, Trainer Prodigy](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Wayta,%20Trainer%20Prodigy) [d86] targeting [Imperial Ceratops](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Imperial%20Ceratops) [eb5][Ripjaw Raptor](file:///C:/Users/Jim/IdeaProjects/mage/Mage.Client/gamelogs/20240105_151800_Commander%20Free%20For%20All_1.html#Ripjaw%20Raptor) [937]
3:19 PM, T1.DA: Human skip attack
3:19 PM, T1.DA: Human has conceded.
3:19 PM, T1.DA: Human has lost the game.
3:19 PM, T1.DA: Computer has won the game
```

Reply:
That is weird.

Looks like it might be a bug in `FightTargetsEffect` - it's unnecessarily attempting to get the card associated with the ability, which doesn't exist for a token, and null checking it.

https://github.com/magefree/mage/blob/f75b1c9f0aa8f17c6df4e2570c39739a9a2691b7/Mage/src/main/java/mage/abilities/effects/common/FightTargetsEffect.java#L35-L38

Reply:
Interesting, i didn't know that tokens arent considered cards